File: built-in plugins dsp/mb-mix-headphone.cpp
/**
 * MB Headphone Mix
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Headphone mixing correction with crossfeed and room simulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_HEADPHONE_H
#define MB_MIX_HEADPHONE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixHeadphone : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-headphone";
    static constexpr const char* PLUGIN_NAME    = "MB Headphone Mix";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float crossfeed = 0.3f;  // range [0, 1]
    float room = 0.4f;  // range [0, 1]
    float bass = 0.3f;  // range [0, 1]
    float output = 0.8f;  // range [0, 1]
    };

    MbMixHeadphone() = default;
    ~MbMixHeadphone() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.crossfeed = std::clamp(params.crossfeed, 0f, 1f);
        params.room = std::clamp(params.room, 0f, 1f);
        params.bass = std::clamp(params.bass, 0f, 1f);
        params.output = std::clamp(params.output, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Headphone Mix
        return input;
    }
};

#endif // MB_MIX_HEADPHONE_H

File: built-in plugins dsp/mb-mix-headphone.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixHeadphonePlugin: PluginDefinition = { id: 'mb-mix-headphone', slug: 'mb-mix-headphone', name: 'MB Headphone Mix', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'Headphone mixing correction with crossfeed and room simulation', author: 'Max Booster', grade: 'A', parameters: [{ id: 'crossfeed', name: 'Crossfeed', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'room', name: 'Room Sim', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }, { id: 'bass', name: 'Bass Comp', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'output', name: 'Output', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { crossfeed: 0.3, room: 0.4, bass: 0.3, output: 0.8 } };

export default MbMixHeadphonePlugin;
